### PR TITLE
Get lang files via FileProviders (and support Razor Class Libraries) 

### PR DESCRIPTION
**Pull request by KevinJump**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

Issue #11986  

Uses the WebHost FileProviders to get the files
this means they don't have to be in the physical root of the site and can be in other places,
so Razor class libraries which use a virtual path provider can be loaded along side files on the site

For testing - i have checked : 

- [x] physical path `/App_Plugins/plugin-name/lang/` 
- [x] physical path `/config/lang/en-us.user.xml`
- [x] razor class libraries :
    - [x]  during dev (linked project so files are in `/path-to-project/wwwroot/lang`
    - [x]  referenced via a nuget file `%userprofile%/.nuget/package/version/staticassets/lang`
 - [x] published site
   - [x]  razor files are physically on site `/wwwroot/app_plugins/plugin-name/lang`

# How to test. 

## Testing on a site in main solution: 
- Place a lang xml file (like the example below) in to the following locations 
    - /App_Plugins/Test/lang/en-us.xml
    - /config/lang/en-us.user.xml
    - /wwwroot/app_plugins/text/lang/en-us.xml

when you load the back office you shoud see the values in the `LocalizedText` file returned by umbraco (viewed via the browser console)
    
## Testing on a new project 
_Building the nuget packages for Umbraco and then starting a new site with dotnet new / package-ref_

- check the tests as above 

To check razor class libraries (RCL) work : 
- create a new Razor Class library project - which has the language file in `wwwroot/lang`, 
- add `<StaticWebAssetBasePath>App_Plugins/RazorPackageTest</StaticWebAssetBasePath>` to the RCL's .csproj file.
- add a project reference to the RCL in your website project. 

### During development (package is linked files are served out of the local nuget cache)
- when loading the umbraco back office the changes from the xml file will be visible (again in LocalizedText). 

### When a site is published (package files are put into the project folder structure). 
- Publish the site. 
- the published site (default `bin/debug/framework/published`) will have the files from the RCL in the `wwwroot/app_plugins/pluginname` folder. 

_Sample lang file for testing_
```xml
<?xml version="1.0" encoding="utf-8" standalone="yes"?>
<language alias="en" intName="English (US)" localName="English (US)" lcid="" culture="en-US">
	<creator>
		<name>Umbraco</name>
		<link>https://umbraco.com</link>
	</creator>
	<area alias="custom">
		<key alias="testHello">Test Hello string</key>
	</area>
</language>
```

### CODE REVIEW

**Comment by KevinJump** [comment]
realised testing this one it a little bit involved, so updated the note to include a bit more detail,

**Comment by bergmania** [comment]
Hi @KevinJump .. Thanks for this PR. I will add it to our next sprint to look into it.

This is something we had on our list already, so we agree it is a good idea :)

One thing that I'm unsure about is whether it makes sense to load language files from embedded resources instead of an RCL. I see no reason these files are accessible from a browser.

**Comment by KevinJump** [comment]
hi @bergmania yeah - I agree, you could even use .resx files! 

based on the discussion with @p-m-j  this is the 'simple' answer that mimics the current behaviour. From a package dev point of view nothing changes, except how you build the project.

I think anything more in depth would need the `LocalizedTextService` to be refactored quite a bit as it is fairly dependent on the files and current XML structure.

**Comment by Zeegaan** [comment]
Thank you for this @KevinJump  and it tests out great 🚀 

Code looks good to me 👓 

As v9/contrib is now no longer used, I have pointed your branch directly to the v10/dev branch and will merge it when tests are happy 👍

File: src/Umbraco.Web.BackOffice/DependencyInjection/UmbracoBuilder.LocalizedText.cs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.FileProviders;
+
+using Umbraco.Cms.Core.DependencyInjection;
+using Umbraco.Cms.Core.Models;
+using Umbraco.Cms.Core.Routing;
+using Umbraco.Cms.Core.Services.Implement;
+
+using IHostingEnvironment = Umbraco.Cms.Core.Hosting.IHostingEnvironment;
```
Comment (Zeegaan) [inline src/Umbraco.Web.BackOffice/DependencyInjection/UmbracoBuilder.LocalizedText.cs:14]:
```suggestion
using System.Collections.Generic;
using System.IO;
using System.Linq;

using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.FileProviders;

using Umbraco.Cms.Core.DependencyInjection;
using Umbraco.Cms.Core.Routing;
using Umbraco.Cms.Core.Services;
```

**Comment by Zeegaan** [comment]
I swear I had this working when testing 🤔 
But I can sadly reproduce @AndyButland's problem :(

**Comment by Zeegaan** [comment]
The lang file does show up when publishing, so i suspect its just the `GetPluginLanguageFileSources` method that needs tinkering with
image

**Comment by KevinJump** [comment]
Guessing we should have a new PR for this one ? 

I have some working code at the moment, my LINQ skills are not hot enough to compress this down - but something like this . 

```cs
private IEnumerable<LocalizedTextServiceSupplementaryFileSource> GetPluginLanguageFileSources(IFileProvider fileProvider, string folder, bool overwrite)
{
    var fileSources = new List<LocalizedTextServiceSupplementaryFileSource>();

    var pluginFolders = fileProvider.GetDirectoryContents(folder)
            .Where(x => x.IsDirectory).ToList();

    foreach(var pluginFolder in pluginFolders)
    {
        // get the full virtual path for the plugin folder
        var pluginFolderPath = WebPath.Combine(folder, pluginFolder.Name);

        // get any lang folders in this plugin                 
        var langFolders = fileProvider.GetDirectoryContents(pluginFolderPath)
            .Where(x => x.IsDirectory && x.Name.InvariantEquals("lang"));

        // loop through the lang folder(s)
        //  - there could be multiple on case sensitive file system
        foreach(var langFolder in langFolders)
        {
            // get the full 'virtual' path of the lang folder 
            var langFolderPath = WebPath.Combine(pluginFolderPath, langFolder.Name);

            // request all the files out of the path, these will have physicalPath set.
            var files = fileProvider.GetDirectoryContents(langFolderPath)
                .Where(x => x.Name.InvariantEndsWith(".xml"))
                .Select(x => new FileInfo(x.PhysicalPath))
                .Select(x => new LocalizedTextServiceSupplementaryFileSource(x, overwrite))
                .ToList();
            
            // add any to our results 
            if (files.Count > 0) 
            { 
                fileSources.AddRange(files);
            }
        }
    }

    return fileSources;
}
```

essentially once you ask for the files explicitly from the folder (e.g `/app_plugins/TestRazorClassLibrary/lang` then you do get the physical path, and you can continue with the file path stuff.

**Comment by KevinJump** [comment]
I also think this now doesn't need to be called twice (once with the `ContentRootFileProvider `and once with the `WebRootFileProvider `as the WebRootFileProvider now also returns the files in the local app_plugins folder ?

**Comment by KevinJump** [comment]
I've put a new PR in for this - and tested it and it appears to work. 🤞

**Comment by AndyButland** [comment]
Amazing work @KevinJump!... thanks very much for the quick look into this.  Will test out the new PR.

**Pull request merged**